data/src/main/java/com/linkedin/data/schema/validation/ValidateDataAgainstSchema.java

```diff
@@ -505,11 +505,12 @@ else if (object instanceof DataMap)
       {
         // Pegasus mode
         DataMap map = (DataMap) object;
-        if (map.size() != 1)
+        // we allow empty union
+        if (map.size() > 1)
```

Comment:
This will allow empty unions even if they were not caused by a projection. 

I was thinking we should allow only when the union has projection applied. One option is to add a custom property (e.g, "$projectedUnion" = "true" ) when we generate the schema from projection and use that to allow empty union here.
What do you think?

Reply:
I thought about it when I was working on it (because it was written in the ticket). 
After making the change, I did some manual integration test and when there is a projection, there will only contain the projected field. I think the projection happens after validation. I think the projection still comes to the picture.

I may have probably misunderstood your idea, feel free to comment

Reply:
The other thought on it was that to introduce projection into validation seems like bound to make it a miss..but that was probably me.

Reply:
AFAIK this is the flow today:

Request -> 1. Validation Filter (create subset schema from projection) -> 2. Resource -> 3. Apply Projection -> 4. Validation Filter (validate using projection schema)

So we can add a flag in the UnionDataSchema created in step 1 using the projection. And then use that flag in the check done in step 4. It is slightly hacky as we are using a property. A cleaner approach would be to add a flag in the union schema to say if empty union data is allowed.